core/src/cgroup.rs

```diff
@@ -1,238 +1,392 @@
+#[derive(Debug)]
 pub struct CGroup {
     path: PathBuf,
 }
 
 impl CGroup {
-    const MEMBER_FILE: &'static str = "cgroup.procs";
-    const FREEZE_FILE: &'static str = "cgroup.freeze";
-    const EVENTS_FILE: &'static str = "cgroup.events";
-
-    /// Returns the first mount point on the host filesystem of the cgroup that this process belongs to
-    pub fn mount_point() -> TypedResult<PathBuf> {
-        procfs::process::Process::myself()
-            .typ(SystemError::Panic)?
-            .mountinfo()
-            .typ(SystemError::Panic)?
-            .iter()
-            .find(|m| m.fs_type.eq("cgroup2")) // TODO A process can have several cgroup mounts
-            .ok_or_else(|| anyhow!("no cgroup2 mount found"))
-            .typ(SystemError::Panic)
-            .map(|m| m.mount_point.clone())
-    }
-
-    /// Create a cgroup inside another one
-    // TODO: Switch to a tree data structure for cgroups.
-    pub fn new<P: AsRef<Path>>(parent: P, child_name: &str) -> TypedResult<Self> {
-        let path = PathBuf::from(parent.as_ref()).join(child_name);
-        //trace!("New CGroup: {path:?}");
+    /// Creates a new cgroup as the root of a sub-tree
+    ///
+    /// path must be the path of an already existing cgroup
+    pub fn new_root<P: AsRef<Path>>(path: P, name: &str) -> anyhow::Result<Self> {
+        let path = PathBuf::from(path.as_ref()).join(name);
+
+        // There is no need to check if the path already exists, as create_dir()
+        // will fail under this circumstance.
+        fs::create_dir(&path)?;
+
+        Self::import_root(&path)
+    }
+
+    /// Imports an already existing cgroup as the root of a sub-tree
+    pub fn import_root<P: AsRef<Path>>(path: P) -> anyhow::Result<Self> {
+        let path = PathBuf::from(path.as_ref());
+
+        if !is_cgroup(&path)? {
+            bail!("{} is not a valid cgroup", path.to_str().unwrap());
+        }
+
+        Ok(CGroup {
+            path: path.to_path_buf(),
+        })
+    }
+
+    /// Creates a sub-cgroup inside this one
+    pub fn new(&self, name: &str) -> anyhow::Result<Self> {
+        Self::new_root(&self.path, name)
+    }
+
+    /// Moves a process to this cgroup
+    pub fn mv(&self, pid: Pid) -> anyhow::Result<()> {
+        if !is_cgroup(&self.path)? {
+            bail!("{} is not a valid cgroup", self.path.to_str().unwrap());
+        }
+
+        fs::write(self.path.join("cgroup.procs"), pid.to_string())?;
+        Ok(())
+    }
+
+    /// Returns all PIDs associated with this cgroup
+    pub fn get_pids(&self) -> anyhow::Result<Vec<Pid>> {
+        if !is_cgroup(&self.path)? {
+            bail!("{} is not a valid cgroup", self.path.to_str().unwrap());
+        }
 
+        // TODO: "Rustify" this
+        let data = fs::read(self.path.join("cgroup.procs"))?;
+        let mut pids: Vec<Pid> = Vec::new();
+        for line in data.lines() {
+            let line = line?;
+            pids.push(Pid::from_raw(line.parse()?));
+        }
```

Comment:
In Rust iterators are really powerful.  
Maybe give [this Blog post about iterators](https://hermanradtke.com/2015/06/22/effectively-using-iterators-in-rust.html/) a read